### PR TITLE
Implement megatron-aware perplexity in torchmetrics

### PR DESCRIPTION
**Pull request by sichu2023**
## Summary
Replace `PerplexityLoggingCallback` with parallelism-aware `MegatronPerplexityMetric`. Pipeline parallel is not supported yet.

## Usage
Added perplexity logging in argparse.
```python
python sub-packages/bionemo-esm2/src/bionemo/esm2/scripts/train_esm2.py --log-train-ppl --log-val-ppl
```

## Testing
Added under `sub-packages/bionemo-llm/tests/bionemo/llm/test_lightning.py`

Tests for these changes can be run via:
```shell
pytest -v sub-packages/bionemo-llm/tests/bionemo/llm/test_lightning.py
```

### CODE REVIEW

**Comment by sichu2023** [comment]
Previous @pstjohn 's effort
https://github.com/NVIDIA/bionemo-framework/pull/382

File: sub-packages/bionemo-llm/src/bionemo/llm/lightning.py [hunk 1]
```diff
@@ -211,6 +212,34 @@ def predict_loss_reduction(self) -> PassthroughLossReduction:
 """
 
 
+class MegatronPerplexityMetric(torchmetrics.text.Perplexity):
+    def __init__(self, *args, **kwargs):
+        if parallel_state.get_context_parallel_world_size() > 1:
+            raise NotImplementedError(f"{self.__class__} does not support context parallelism yet.")
+
+        self.cross_entropy_loss_fusion = kwargs.pop("cross_entropy_loss_fusion", False)
```
Comment (pstjohn) [inline sub-packages/bionemo-llm/src/bionemo/llm/lightning.py:220]: My only concern here is that this is a perplexity-specific implementation. Is there no way we can do this without a call to `unreduced_token_loss_fn`?

Comment (sichu2023) [reply sub-packages/bionemo-llm/src/bionemo/llm/lightning.py:220]: `unreduced_token_loss_fn` is the only tp-aware nll functions we have. Replacing would just be reinventing the wheel.

Comment (sichu2023) [reply sub-packages/bionemo-llm/src/bionemo/llm/lightning.py:220]: My major concern is a lack of robust testing under parallelism in unittest. We don't have that infrastructure yet.

File: sub-packages/bionemo-esm2/src/bionemo/esm2/scripts/train_esm2.py [hunk 2]
```diff
@@ -213,7 +221,13 @@ def main(
         log_every_n_steps=log_every_n_steps,
         num_nodes=num_nodes,
         callbacks=callbacks,
-        plugins=nl.MegatronMixedPrecision(precision=precision),
+        plugins=nl.MegatronMixedPrecision(
+            precision=precision,
+            params_dtype=get_autocast_dtype(precision),
+            pipeline_dtype=get_autocast_dtype(precision),
+            grad_reduce_in_fp32=False,
```
Comment (jstjohn) [inline sub-packages/bionemo-esm2/src/bionemo/esm2/scripts/train_esm2.py:228]: Same here re reduce in fp32. I think the ddp setting takes precedence over this but I'm not sure.

File: sub-packages/bionemo-llm/src/bionemo/llm/lightning.py [hunk 3]
```diff
@@ -211,6 +212,34 @@ def predict_loss_reduction(self) -> PassthroughLossReduction:
...
+        super().__init__(*args, **kwargs)
+
+    def update(self, preds: torch.Tensor, target: torch.Tensor) -> None:
+        """Update state with predictions and targets under tensor parallelism."""
+        unreduced_token_loss = unreduced_token_loss_fn(  # TP-aware log prob function
+            preds.clone(),
+            target.clone(),
+            cross_entropy_loss_fusion=self.cross_entropy_loss_fusion,
```
Comment (jstjohn) [inline sub-packages/bionemo-llm/src/bionemo/llm/lightning.py:228]: nice

Comment (sichu2023) [reply sub-packages/bionemo-llm/src/bionemo/llm/lightning.py:228]: I am planning to create another PR on configurable loss and metric. How about I bundle this into this other PR?

File: sub-packages/bionemo-esm2/src/bionemo/esm2/scripts/train_esm2.py (hunk 2 above)
Comment (sichu2023) [reply sub-packages/bionemo-esm2/src/bionemo/esm2/scripts/train_esm2.py:228]: Move to https://github.com/NVIDIA/bionemo-framework/pull/520

File: sub-packages/bionemo-esm2/src/bionemo/esm2/scripts/train_esm2.py [hunk 4]
```diff
@@ -163,10 +164,17 @@ def main(
...
         pipeline_model_parallel_size=pipeline_model_parallel_size,
-        ddp="megatron",
+        pipeline_dtype=get_autocast_dtype(precision),
+        ddp=DistributedDataParallelConfig(
+            check_for_nan_in_grad=True,
+            overlap_grad_reduce=True,
+            overlap_param_gather=True,
+            average_in_collective=True,
+            use_distributed_optimizer=True,
+        ),
```
[the remark this answers is not in the document]
Comment (sichu2023) [reply sub-packages/bionemo-esm2/src/bionemo/esm2/scripts/train_esm2.py:174]: Move to https://github.com/NVIDIA/bionemo-framework/pull/520

**Comment by sichu2023** [comment]
Exponential Validation Loss vs Validation Perplexity
Test run on single device without parallelism.

**Comment by sichu2023** [comment]
On multi-device DP
Exponential of Validation Loss and Validation PPL over Time

**Comment by sichu2023** [comment]
And on multi-device TP x DP.
